packages/dds/tree/src/simple-tree/toMapTree.ts

```diff
@@ -289,21 +311,26 @@ function objectToMapTree(data: InsertableContent, schema: TreeNodeSchema): MapTr
 	const fields = new Map<FieldKey, MapTree[]>();
 
 	// Filter keys to only those that are strings - our trees do not support symbol or numeric property keys
-	const keys = Reflect.ownKeys(data).filter((key) => typeof key === "string") as FieldKey[];
+	const keys = Reflect.ownKeys(data).filter((key) => typeof key === "string") as string[];
 
 	for (const viewKey of keys) {
-		const fieldValue = (data as Record<FieldKey, InsertableContent>)[viewKey];
-
-		// Omit undefined record entries - an entry with an undefined key is equivalent to no entry
-		if (fieldValue !== undefined) {
-			const fieldSchema = getObjectFieldSchema(schema, viewKey);
-			const mappedChildTree = nodeDataToMapTree(fieldValue, fieldSchema.allowedTypeSet);
-			const flexKey: FieldKey = brand(getStoredKey(viewKey, fieldSchema));
-
-			// Note: SchemaFactory validates this at schema creation time, with a user-friendly error.
-			// So we don't expect to hit this, and if we do it is likely an internal bug.
-			assert(!fields.has(flexKey), 0x925 /* Keys must not be duplicated */);
-			fields.set(flexKey, [mappedChildTree]);
+		const fieldValue = (data as Record<string, InsertableContent>)[viewKey];
+		const fieldSchema = getObjectFieldSchema(schema, viewKey);
+		setFieldValue(fields, fieldValue, fieldSchema, nodeKeyManager, viewKey);
+	}
+
+	// Loop through field keys without data, and assign value from its default provider.
+	for (const [key, fieldSchema] of Object.entries(
+		schema.info as Record<string, ImplicitFieldSchema>,
+	)) {
+		if (!keys.includes(key)) {
```

Comment:
I changed the structure of the function a bit (as suggested in a comment below), so I am now checking if `keys.includes(key)` first, and otherwise calling the default provider to get the value.

It looks like `keys` is a subset of the keys in `data` since we only want the keys that are strings, so would it be better to leave it using the `keys` instead of `data`?

Reply:
Right - I don't think this is relevant anymore given the structural change.